arc/src/main/java/org/apache/james/arc/ARCChainValidator.java

```diff
@@ -0,0 +1,164 @@
+ * This class provides methods to validate the ARC chain by checking the structure,
+ * verifying ARC-Message-Signature and ARC-Seal headers, and ensuring the integrity
+ * of previous ARC hops. It uses DNS records and cryptographic verification to
+ * ensure the authenticity of the ARC chain.
+ * </p>
+ */
+public class ARCChainValidator {
+    public static final String ARC_MESSAGE_SIGNATURE = "ARC-Message-Signature";
+    public static final String ARC_SEAL = "ARC-Seal";
+    private static final String SHA256_RSA = "SHA256withRSA";
+    private final Pattern INST_RGX_PATTERN = Pattern.compile("i=([0-9]+)");
+    private final PublicKeyRetrieverArc _keyRecordRetriever;
+
+    public ARCChainValidator(PublicKeyRetrieverArc keyRecordRetriever) {
+        this._keyRecordRetriever = keyRecordRetriever;
+    }
+
+    public String validateArcChain(Message message) {
+
+        Header messageHeaders = message.getHeader();
+        int curInstance = getCurrentInstance(messageHeaders);  // Incremented by 1
+
+        if (curInstance == 1) { //we are the first ARC Hop and there is no previous ARC hops in the chain to validate
+            return "none";
+        }
+        else if (curInstance > 51) { // Not allowed to be > 50
+            return "fail";
+        }
+        else { // there are previous ARC hops that need to be validated
+            return validatePreviousArcHops(message, messageHeaders, curInstance);
+        }
+    }
+
+    private String validatePreviousArcHops(Message message, Header messageHeaders, int myInstance) {
+        ARCVerifier arcVerifier = new ARCVerifier(_keyRecordRetriever);
+        Map<Integer, List<Field>> arcHeadersByI = arcVerifier.getArcHeadersByI(messageHeaders.getFields());
+        int numArcInstances = myInstance -1;
+        boolean isArcSetStructureOK = arcVerifier.validateArcSetStructure(arcHeadersByI);
+        if (!isArcSetStructureOK) {
+            return  "fail";
+        }
+
+        Set<Field> prevArcSet;
+        prevArcSet = arcVerifier.extractArcSet(messageHeaders, numArcInstances);
+        if  (prevArcSet != null) {
+            boolean amsOk = checkArcAms (prevArcSet, message, arcVerifier);
+            boolean asOk = checkArcSeal (messageHeaders.getFields(), numArcInstances, arcVerifier);
+            if (amsOk && asOk) {
+                return "pass";
+            }
+        }
+        return  "fail";
+    }
+
+    private boolean checkArcAms(Set<Field> prevArcSet, Message message, ARCVerifier arcVerifier){
+        boolean retVal = false;
+
+        Field amsHeader = prevArcSet.stream()
+                .filter(f -> f.getName().equalsIgnoreCase(ARC_MESSAGE_SIGNATURE))
+                .findFirst().orElse(null);
+        if (amsHeader == null) return retVal;
+
+        String txtDnsRecord = arcVerifier.getTxtDnsRecordByField(amsHeader);
+        if (txtDnsRecord == null) return retVal;
+
+        retVal = arcVerifier.verifyAms(amsHeader, message, txtDnsRecord);
+
+        return retVal;
```

Comment:
```suggestion
        return arcVerifier.verifyAms(amsHeader, message, txtDnsRecord);
```

And we get rid of retVal?